feature/global_ens_priorities_456

malloryprow opened the pull request:
Since global_ens is such a large component and I am not the developer of this code,  we are going to iteratively work through this PR, so the usual PR information will be found in the comments for each iteration.

This PR can be merged once all iterations are completed.

- [x] Iteration 1: global_ens wave: prep, stats, plots
- [x] Iteration 2: global_ens atmos & headline: prep
- [x] Iteration 3: global_ens atmos & headline: stats

Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
**Set-up**
> 1. Clone my fork and checkout branch feature/global_ens_priorities_456
> 2. cd sorc
> 3. ./build-fix-link.sh
> 4. ./build.sh

- [ ] Please complete this pull request review by **11/09/2023**.</br>

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

malloryprow committed:
Replace CYC with new, appropriate variables names

Refs: #5

malloryprow committed:
Replace cycles with new, appropriate variables names

Refs: #5

malloryprow committed:
Merge branch 'develop' into feature/global_ens_cyc_change

Refs: #5

malloryprow committed:
Replace cycle with new, appropriate variables names

Refs: #5

malloryprow committed:
Replace cyc with new, appropriate variables names

Refs: #5

malloryprow committed:
Merge branch 'develop' into feature/global_ens_cyc_change

Refs: #5

malloryprow committed:
Add vhr in J-Jobs

Refs: #5

malloryprow committed:
Add vhr in wave dev drivers scripts

Refs: #8

malloryprow committed:
Merge branch 'develop' into feature/global_ens_cyc_change

Refs: #5

malloryprow committed:
Merge branch 'develop' into feature/global_ens_cyc_change

Refs: #5

malloryprow committed:
Use ihour instead of vhour

Refs: #5

malloryprow committed:
Remove unused environment variables

Refs: #5

malloryprow committed:
Merge branch 'develop' into feature/global_ens_cyc_change

Refs: #5

malloryprow committed:
Update script name

Refs: #5

malloryprow committed:
Add evs_ver and evs_ver_2d

Refs: #5

malloryprow committed:
Remove unused global_ens files in metplus_config

Refs: #5

malloryprow committed:
Remove more unused global_ens files in metplus_config

Refs: #5

malloryprow committed:
Remove MORE unused global_ens files in metplus_config

Refs: #5

malloryprow committed:
Add LOG_METPLUS as empty string

Refs: #5

malloryprow committed:
Cat the plotting log files

Refs: #5

malloryprow committed:
Add set -x

Refs: #5

malloryprow committed:
Remove unused ush scripts

Refs: #5

malloryprow committed:
Remove exevs_global_ens_aerosol_stats.sh

Refs: #5

malloryprow committed:
Add interpreter at the top of scripts

Refs: #5

malloryprow committed:
Remove METplus machine.conf settings

Refs: #5

malloryprow committed:
Remove hard-coded paths

Refs: #5

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

Refs: #5

malloryprow committed:
List all possible missing CCPA files in missing email message

Refs: #5

malloryprow committed:
Fix wave plots legends

Refs: #5

malloryprow committed:
Remove jlogfile and postmsg usage

Refs: #5

malloryprow committed:
Add restrictions for rstprod data

Refs: #5

malloryprow committed:
Remove more hard-coded paths

Refs: #5

malloryprow committed:
Final EE2 change for global_ens wave

malloryprow committed:
Save atmos prep changes

malloryprow commented:
### Iteration 1: global_ens wave: prep, stats, plots

1. Rename variables containing "CYC", "cyc", "cycle", "cycles" or remove if the instance isn't needed
2. Add evs_ver and evs_ver_2d and update COMIN and COMOUT paths in dev/drivers/scripts
3. Remove use of jlogfile and postmsg
4. Remove hard-coded paths
5. Clean up "LOUD" usage
6. Use cpreq (along with file checking)
7. Use prod_util's ndate for date mainpulation
8. Add use of err_chk and err_exit
9. Cat the plotting log files to see contents in main log file (for plots step)
10. Use "LOG_METPLUS =" to MET output goes to main log file (for prep and stats)
11. Add restrictions on rstprod data 
12. Add WARNING to non-fatal logging statements
13. Add python and shell interpreter at top of scripts
14. Add "set -x"
15. Fix plot legend


**Describe testing already performed for this Pull Request:**
> I ran the prep, stats, and plots jobs and compared the output to the parallel. All steps matched the parallel.

**Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:**

> 1. Run dev/drivers/scripts/prep/global_ens/jevs_global_ens_wave_grid2obs_prep.sh
> 2. Run dev/drivers/scripts/stats/global_ens/jevs_global_ens_wave_grid2obs_stats.sh
> 3. Run dev/drivers/scripts/plots/global_ens/jevs_global_ens_wave_grid2obs_plots.sh

**Has the code been checked to ensure that no errors occur during the execution?** Yes

**Do these updates/additions include sufficient testing updates?** Yes

ShelleyMelchior-NOAA commented:
wave prep is underway. So far the grib2 files match. Just waiting on the nc files.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/global_ens/wave.20231101`
logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_wave_grid2obs_prep.o107968298`

ShelleyMelchior-NOAA commented:
wave stats has completed.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/stats/global_ens/`
logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_wave_grid2obs_stats.o107969009`

ShelleyMelchior-NOAA commented:
@malloryprow , since vhr is defaulted to 00 in the wave J-jobs, can we remove it's declaration in the driver scripts?

ShelleyMelchior-NOAA commented:
wave plots has completed.
COMOUT=`/lfs/h2/emc/ptmp/shelley.melchior/evs/v1.0/plots/global_ens/wave.20231101`
logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_wave_grid2obs_plots.o107969498`

malloryprow commented:
Reviewing!

StevenSimon-NOAA commented:
> wave prep is underway. So far the grib2 files match. Just waiting on the nc files. COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/global_ens/wave.20231101` logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_wave_grid2obs_prep.o107968298`

All wave prep files are fully accounted for by quantity. There are no apparent errors in the accompanying logfile.

StevenSimon-NOAA commented:
> wave stats has completed. COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/stats/global_ens/` logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_wave_grid2obs_stats.o107969009`

All wave stats files are fully accounted for by quantity. There are no apparent errors in the accompanying logfile.

StevenSimon-NOAA commented:
> wave plots has completed. COMOUT=`/lfs/h2/emc/ptmp/shelley.melchior/evs/v1.0/plots/global_ens/wave.20231101` logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_wave_grid2obs_plots.o107969498`

All wave tar files and contained plot pngs are fully accounted for by quantity. There are no apparent errors in the accompanying logfile.

StevenSimon-NOAA commented:
> Reviewing!

I saw on the Slack Channel that the PR was submitted and started doing some of the QC review. I hope I did not step on your toes here, since I just saw your comment on the PR thread.

malloryprow commented:
Everything checks out for iteration 1. The output matches the parallel and is what is expected. I grep'd the log files for "bad" words and didn't find any. I would say iteration 1 is good.

Iteration 2 will be the global_ens atmos prep. Some of the changes I made were making the job run a lot longer than how it normally runs in the parallel (I was looping through by forecast hours checking for the existence of the file and then running METplus for that forecast hour, where as before METplus was run for a bunch of forecast hours and not getting for the existence of the files). I need to play around and see if I can do anything to speed this up. I'll let you know when we can work on the next round of testing for this PR!

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

malloryprow commented:
> > Reviewing!
> 
> I saw on the Slack Channel that the PR was submitted and started doing some of the QC review. I hope I did not step on your toes here, since I just saw your comment on the PR thread.

The more eyes the better!

ShelleyMelchior-NOAA commented:
> @malloryprow , since vhr is defaulted to 00 in the wave J-jobs, can we remove it's declaration in the driver scripts?

Just wanted to bump this in case it got missed, @malloryprow .

malloryprow commented:
Ah yes! So this is the one of the lines you are talking about?

https://github.com/malloryprow/EVS/blob/d7cd3dc5642637e55b055ad321b53c4d37604d1a/dev/drivers/scripts/prep/global_ens/jevs_global_ens_wave_grid2obs_prep.sh#L27

ShelleyMelchior-NOAA commented:
> Ah yes! So this is the one of the lines you are talking about?
> 
> https://github.com/malloryprow/EVS/blob/d7cd3dc5642637e55b055ad321b53c4d37604d1a/dev/drivers/scripts/prep/global_ens/jevs_global_ens_wave_grid2obs_prep.sh#L27

Yep!

malloryprow committed:
Remove vhr

malloryprow committed:
Clean up empty lines

malloryprow committed:
Use array to gather all avaiable fhrs to run METplus once

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

malloryprow committed:
Fix resources

malloryprow committed:
Used mkdir -p

malloryprow committed:
Rework headline stats

malloryprow committed:
Move some SENDCOM statements

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

malloryprow commented:
### Iteration 2: global_ens atmos and headline: prep

1. Adjust resources used
2. Rename variables containing "CYC", "cyc", "cycle", "cycles" or remove if the instance isn't needed
3. Add evs_ver and evs_ver_2d and update COMIN and COMOUT paths in dev/drivers/scripts
4. Remove use of jlogfile and postmsg
5. Use "LOG_METPLUS =" to MET output goes to main log file
6. Remove METplus settings defined individual global_ens METplus configs that are defined in machine.conf
7. Remove unused files
8. Add use of err_chk and err_exit
9. Add checking of files
10. Write files to DATA instead of COMOUT
11. Use cpreq (along with file checking)
12. Add restrictions on rstprod data
13. General code cleanup

Describe testing already performed for this Pull Request:

> I ran the prep atmos and headline jobs. Due to misplaced SENDCOM copy statements, the new COMOUT directories for atmos has more files. Logs were checked for the following: “error”, “Error”, “ERROR”, “fail”, “Fail”, “FAIL”, “Syntax”, “argument expected”, “No such file”, “exceeded”. None were found.

Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:

> 1. Run dev/drivers/scripts/prep/global_ens/jevs_global_ens_atmos_prep.sh and dev/drivers/scripts/prep/global_ens/jevs_global_ens_naefs_atmos_prep.sh.
> 2. Run dev/drivers/scripts/prep/global_ens/jevs_global_ens_headline_prep.sh. This job uses output from dev/drivers/scripts/prep/global_ens/jevs_global_ens_atmos_prep.sh. In the parallel, this job starts an hour after dev/drivers/scripts/prep/global_ens/jevs_global_ens_atmos_prep.sh even though the jobs runs longer than that. 

Has the code been checked to ensure that no errors occur during the execution? **Yes**

Do these updates/additions include sufficient testing updates? **Yes**

ShelleyMelchior-NOAA commented:
The prep processing for `jevs_global_ens_naefs_atmos_prep.sh` and `jevs_global_ens_atmos_prep.sh` has been submitted. The naefs prep processing has completed, but the gefs processing is still underway.
```
Job ID                         Username        Queue           Jobname         SessID   NDS  TSK   Memory Time  S Time
------------------------------ --------------- --------------- --------------- -------- ---- ----- ------ ----- - -----
108259991.dbqs01               shelley.melchi* dev             jevs_global_en*   135057    1    83  350gb 06:00 R 00:04
```
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/global_ens/atmos.20231104`
logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_*_prep.o10825999*`

I will start the headline prep job around 3:00p.

StevenSimon-NOAA commented:
> The prep processing for `jevs_global_ens_naefs_atmos_prep.sh` and `jevs_global_ens_atmos_prep.sh` has been submitted. The naefs prep processing has completed, but the gefs processing is still underway.
> 
> ```
> Job ID                         Username        Queue           Jobname         SessID   NDS  TSK   Memory Time  S Time
> ------------------------------ --------------- --------------- --------------- -------- ---- ----- ------ ----- - -----
> 108259991.dbqs01               shelley.melchi* dev             jevs_global_en*   135057    1    83  350gb 06:00 R 00:04
> ```
> 
> COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/global_ens/atmos.20231104` logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_*_prep.o10825999*`
> 
> I will start the headline prep job around 3:00p.

All NAEFS prep files (cmce_bc and gefs_bc) are fully accounted for by quantity. The corresponding logfile shows no errors.

StevenSimon-NOAA commented:
All headline prep files (cmce and gefs) are fully accounted for by quantity. The corresponding logfile shows no errors.

ShelleyMelchior-NOAA commented:
Thanks, Steve! I re-ran the headline prep job just now just to better honor the lineup that Mallory acknowledged in the emc.vpppg crontab.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/global_ens/headline.20231104`
logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_headline_prep.o108264517`

StevenSimon-NOAA commented:
> Thanks, Steve! I re-ran the headline prep job just now just to better honor the lineup that Mallory acknowledged in the emc.vpppg crontab. COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/global_ens/headline.20231104` logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_headline_prep.o108264517`

The headline prep files and logfile look good after the rerun.

malloryprow commented:
Sorry about the confusion with the CCPA data in dev/drivers/scripts/prep/global_ens/jevs_global_ens_atmos_prep.sh. I forgot about that. 

Here is this snippet of the code:

```
if [ $modnam = ccpa ] ; then
  day1=$($NDATE -24 ${vday}12)
  export vday_1=${day1:0:8}
  for ihour in 00 06 12 18 ; do
    if [ -s $COMINccpa/ccpa.${vday}/$ihour/ccpa.t${ihour}z.06h.1p0.conus.gb2 ] ; then
      $WGRIB2 $COMINccpa/ccpa.${vday}/$ihour/ccpa.t${ihour}z.06h.1p0.conus.gb2 -set_grib_type same -new_grid_winds earth -new_grid ncep grid 003  $WORK/ccpa.t${ihour}z.grid3.06h.f00.grib2
      [[ $SENDCOM="YES" ]] && cpreq -v $WORK/ccpa.t${ihour}z.grid3.06h.f00.grib2 ${COMOUTgefs}/ccpa.t${ihour}z.grid3.06h.f00.grib2
    else
        if [ $SENDMAIL = YES ]; then
          export subject="CCPA  Data Missing for EVS ${COMPONENT}"
          echo "Warning:  No CCPA analysis available for ${INITDATE}${ihour}" > mailmsg
          echo "Missing file is $COMINccpa/ccpa.${vday}/$ihour/ccpa.t${ihour}z.06h.1p0.conus.gb2"  >> mailmsg
          echo "Job ID: $jobid" >> mailmsg
          cat mailmsg | mail -s "$subject" $maillist
	fi
    fi 
  done
  export output_base=${WORK}/precip
  export ccpa24=${WORK}/ccpa24
  mkdir -p $ccpa24
  for ihour in 12 ; do
    cpreq -v ${COMOUTgefs}/ccpa.t12z.grid3.06h.f00.grib2 ${WORK}/ccpa24/ccpa1
    cpreq -v ${COMOUTgefs}/ccpa.t06z.grid3.06h.f00.grib2 ${WORK}/ccpa24/ccpa2
    cpreq -v ${COMOUTgefs}/ccpa.t00z.grid3.06h.f00.grib2 ${WORK}/ccpa24/ccpa3
    cpreq -v ${COMOUT}/atmos.${vday_1}/gefs/ccpa.t18z.grid3.06h.f00.grib2 ${WORK}/ccpa24/ccpa4
    if [ -s ${WORK}/ccpa24/ccpa1 ] && [ -s ${WORK}/ccpa24/ccpa2 ] && [ -s ${WORK}/ccpa24/ccpa3 ] && [ -s ${WORK}/ccpa24/ccpa4 ] ; then
       ${METPLUS_PATH}/ush/run_metplus.py -c ${PARMevs}/metplus_config/machine.conf -c ${CONF_PREP}/PcpCombine_obsCCPA24h.conf
       [[ $SENDCOM="YES" ]] && cpreq -v $output_base/ccpa.t12z.grid3.24h.f00.nc $COMOUTgefs/.
    else
       if [ $SENDMAIL = YES ]; then
         export subject="06h CCPA Data Missing for 24h CCPA generation"
         echo "Warning: At least one of ccpa06h files is missing for ${vday}${ihour}" > mailmsg
         echo "Missing file is ${COMOUTgefs}/ccpa.t00z.grid3.06h.f00.grib2, ${COMOUTgefs}/ccpa.t06z.grid3.06h.f00.grib2, ${COMOUTgefs}/ccpa.t12z.grid3.06h.f00.grib2, or ${COMOUT}.${vday_1}/gefs/ccpa.t18z.grid3.06h.f00.grib2"  >> mailmsg
         echo "Job ID: $jobid" >> mailmsg
         cat mailmsg | mail -s "$subject" $maillist
       fi
    fi  
  done
fi
```

I see the copying is coming from COMOUT. I will switch that to COMIN.

malloryprow commented:
I think I can change it from having that dependency on that ${COMOUT}/atmos.${vday_1}/gefs/ccpa.t18z.grid3.06h.f00.grib2 file. Any thoughts for or against?

AliciaBentley-NOAA commented:
@malloryprow I don't think I mind the dependency? Up to you, really! Less work is good too!

malloryprow committed:
Clean up mail message for CCPA data

malloryprow commented:
Okay! I kept it, but cleaned up the mail messaging.

ShelleyMelchior-NOAA commented:
Thanks, Mallory. I will run global_ens_atmos_prep again to see the email message change.

ShelleyMelchior-NOAA commented:
Ok, the re-run is underway.
```
Job ID                         Username        Queue           Jobname         SessID   NDS  TSK   Memory Time  S Time
------------------------------ --------------- --------------- --------------- -------- ---- ----- ------ ----- - -----
108308841.dbqs01               shelley.melchi* dev             jevs_global_en*   141329    1    83  350gb 06:00 R 00:00
```
logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_atmos_prep.o108308841`

StevenSimon-NOAA commented:
Thanks Mallory for addressing the "6hr ccpa file missing for 24hr ccpa generation" error message issue and incorporating it into this PR. Ultimately, we do need the prior valid date's 18z 6hr ccpa file to generate the 24hr ccpa netcdf file, so it is definitely worth retaining that particular dependency here, but formatting it in terms of COMIN as you did in the most recent edits. Once global_ens_atmos_prep runs to completion, I will do a QC review of the generated files and logfile.

ShelleyMelchior-NOAA commented:
> Thanks Mallory for addressing the "6hr ccpa file missing for 24hr ccpa generation" error message issue and incorporating it into this PR. Ultimately, we do need the prior valid date's 18z 6hr ccpa file to generate the 24hr ccpa netcdf file, so it is definitely worth retaining that particular dependency here, but formatting it in terms of COMIN as you did in the most recent edits. Once global_ens_atmos_prep runs to completion, I will do a QC review of the generated files and logfile.

Thanks Steve. It is 53 minutes in so far. You can run this command to keep an eye on it, or just monitor the log file.
`qstat -tu shelley.melchior -w | grep global_en`

malloryprow committed:
Adjust headline stats jobs resources

AliciaBentley-NOAA commented:
@ShelleyMelchior-NOAA @StevenSimon-NOAA @malloryprow Are we still waiting on the rerun of the prep job?

malloryprow commented:
Yeah it takes about 5 hours to run

AliciaBentley-NOAA commented:
@malloryprow Got it! Thanks! And Shelley tells me that you've moved on to working on the global_ens stats code?

malloryprow commented:
Yup tests for headline scores has good. I made the changes for grid2obs and in the process of testing it down and debugging the issues I come across.

StevenSimon-NOAA commented:
I would hold off on testing stats right now. Many of the ecme stats files are not being generated on the Parallel. MetPlus is not running for ecme grid2grid for the past week or so, on and off. It is not immediately apparent in the logfiles as to why this is occurring. This has been occurring for a couple of weeks but also overlaps with a period in which ecme prep files were not being fully generated due to Dogwood slowness and missing validation and forecast data. What I can discern from the ecme grid2grid stats logfile is that some validation data may be missing here.  

Here is a sample logfile I have been digging through this morning from yesterday's run:  /lfs/h2/emc/ptmp/emc.vpppg/output/jevs_global_ens_ecme_atmos_grid2grid_stats.o108231923.

I am at a bit of a loss right now as to what the error could be. The logfile claims that there is missing prep validation data, but that does not appear to be the case when I check the ecme sub dir under the global_ens prep directory. 

+ subject='EC analysis data missing '
+ export subject
+ echo 'Warning: No EC analysis available for 20231105'
+ 1> mailmsg
+ echo Missing file is /lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/prep/global_ens/atmos.20231105/ecme/ecmanl.t00z.grid3.f000.grib2
+ 1>> mailmsg
+ echo 'Job ID: LLjevs_global_ens_ecme_atmos_grid2grid_stats.o152472'
+ 1>> mailmsg
+ mail -s 'EC analysis data missing ' alicia.bentley@noaa.gov,steven.simon@noaa.gov
+ cat mailmsg
+ exit
+ msg='JOB jevs_global_ens_ecme_atmos_grid2grid_stats HAS COMPLETED NORMALLY.'
+ postmsg 'JOB jevs_global_ens_ecme_atmos_grid2grid_stats HAS COMPLETED NORMALLY.'

This may be something you already identified in your debugging, but a heads-up if you have not come across it yet.

ShelleyMelchior-NOAA commented:
The prep job re-run died early.
See line  3768517 of the logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_atmos_prep.o108308841`
I know that isn't the actual problem, but rather in that processing that the the job died. Still tracing ... not sure ... but wanted to drop this here for review.

malloryprow committed:
Update addition expression

malloryprow committed:
Rework atmos sea_ice stats

malloryprow committed:
Remove rogue $

malloryprow commented:
> The prep job re-run died early. See line 3768517 of the logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_atmos_prep.o108308841` I know that isn't the actual problem, but rather in that processing that the the job died. Still tracing ... not sure ... but wanted to drop this here for review.

There was a rogue "$". I have removed it in https://github.com/NOAA-EMC/EVS/pull/332/commits/cc396d187731f724c79e9fe8998ea494c9de7bed.

ShelleyMelchior-NOAA commented:
Ok, I re-submitted the global_ens_atmos_prep job just now.
logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_atmos_prep.o108328368`

malloryprow commented:
Not seeing the syntax error anymore in /lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_atmos_prep.o108328368 :+1:

malloryprow committed:
Rework atmos grid2obs stats

malloryprow committed:
Remove unused file

malloryprow committed:
Rework atmos snowfall stats

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

ShelleyMelchior-NOAA commented:
And the global_ens_atmos_prep process has completed!

StevenSimon-NOAA commented:
> /lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/global_ens/

All cmce, gefs, and osi_saf prep files are fully accounted for by quantity. However, the ecme count is 12,002 files (currently 10,552 files outside of this PR). From what I can see, ecme.ensxx.t12z.grid4.24h.f024.nc and ecme.ensxx.t00z.grid4.24h.f024.nc now exist for both init times (00z and 12z, was previously just 12z). Was that a bug fix in this PR?

malloryprow commented:
> > /lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/global_ens/
> 
> All cmce, gefs, and osi_saf prep files are fully accounted for by quantity. However, the ecme count is 12,002 files (currently 10,552 files outside of this PR). From what I can see, ecme.ensxx.t12z.grid4.24h.f024.nc and ecme.ensxx.t00z.grid4.24h.f024.nc now exist for both init times (00z and 12z, was previously just 12z). Was that a bug fix in this PR?

Correct, @StevenSimon-NOAA,  this was a fix. Below is a snippet of what is currently in develop

```
if [ $modnam = ecme_apcp24h ] ; then

  export cyc
  export mb
  export modelpath=$COMOUTecme

  for cyc in 00 12 ; do
     export output_base=${WORK}/precip/ecme_apcp24h.${cyc}
     for mb in 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40  41 42 43 44 45 46 47 48 49 50  ; do
       if [ -s $COMOUTecme/ecme.ens${mb}.t${cyc}z.grid4_apcp.f024.grib1 ] ; then  
         export lead='24, 36, 48, 60, 72, 84, 96,108, 120, 132, 144, 156, 168, 180, 192,204, 216, 228, 240, 252, 264, 276, 288, 300, 312, 324, 336, 348, 360'
	 ${METPLUS_PATH}/ush/run_metplus.py -c ${PARMevs}/metplus_config/machine.conf -c ${CONF_PREP}/PcpCombine_fcstECME_APCP24h.conf
       fi
     done
  done

     [[ $SENDCOM="YES" ]] && cp ${output_base}/*.nc $COMOUTecme/.
fi
```

What gets sent to COMOUTecme is from output_base, which is a directory that contains information from the cyc loop however the SENDCOM statement is outside of the loop thus only the 12z files were being copied. (NOTE: in this PR cyc variable has been renamed). This was also fixed for the gefs_snow24h cmce_snow24h, and ecme_snow24h files.

ShelleyMelchior-NOAA commented:
Fantastic. This is a good fix. Thanks for the explanation, Mallory.
Summary:
wave -- prep, stats, plots ✔️ 
atmos -- prep ✔️

malloryprow commented:
Yup!

I checked /lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_atmos_prep.o108328368, /lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_headline_prep.o108264517, and for the following: “error”, “Error”, “ERROR”, “FATAL ERROR”, “fail”, “Fail”, “FAIL”, “No such file”, “exceeded”, too many arguments", "command not found", "argument expected", "mv: cannot stat", "exited with code 1". I didn't find anything.

Iteration 2 is ✔️ 👍

malloryprow committed:
Rework atmos grid2grid, precip, sst, and cnv

malloryprow committed:
Remove NAEFS cloud verification, remove F000 TCDC GEFS verification

malloryprow committed:
Bug fixes

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

malloryprow committed:
Add back OPM_NUM_THREADS

malloryprow commented:
### Iteration 3: global_ens atmos & headline: stats

1. Add "set -x"
2. Rename variables containing "CYC", "cyc", "cycle", "cycles" or remove if the instance isn't needed
3. Add evs_ver and evs_ver_2d and update COMIN and COMOUT paths in dev/drivers/scripts
4. Adjust job resources
5. Remove use of jlogfile and postmsg
6. General script cleanup
7. Use "LOG_METPLUS =" to MET output goes to main log file
8. Remove METplus settings defined individual global_ens METplus configs that are defined in machine.conf
9. Remove unused METplus config files
10. Fix GRID_STAT_CLIMO_MEAN_FIELD in precip climo_prob METplus configs to remove MET warning messages
11. Don't run TCDC stats for GEFS at F000
12.  Add use of err_chk and err_exit
13. Fix checking of wrong file for ECMWF for grid2grid stats
14. Add interpreter's at top of scripts
15. Add mail alerts for missing data
17. Use cpreq (along with file checking)
23. Add WARNING to non-fatal logging statements

NOTE: The WMO stats related scripts were not touched as these will not be run in operations.

**Describe testing already performed for this Pull Request:**
> I ran the stats and compared the output to the parallel.

**Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:**

> 1. We will need to run all the scripts in dev/drivers/scripts/stats/global_ens minus the WMO ones.

**Has the code been checked to ensure that no errors occur during the execution?** Yes

**Do these updates/additions include sufficient testing updates?** Yes

malloryprow committed:
Merge branch 'develop' into feature/global_ens_priorities_456

ShelleyMelchior-NOAA commented:
@malloryprow  and @StevenSimon-NOAA , I have kicked off all of the stats processing. All are underway and some are completed. Please review.
logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_*_stats.o93093*`
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/stats/global_ens/gefs.20231111`

StevenSimon-NOAA commented:
> @malloryprow and @StevenSimon-NOAA , I have kicked off all of the stats processing. All are underway and some are completed. Please review. logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_*_stats.o93093*` COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/stats/global_ens/gefs.20231111`

There are a few errors that we need to navigate through. But let's start with the precip stats drivers first. The ecme precip stats job has a syntax error around line 40000. The cmce precip stats job has a bunch of LNET errors in debugging section, but it is not immediately clear what the source is. The naefs precip stats job has a similar syntax error as ecme prcip stats. The gefs precip stats job has the same unknown LNET error as the cmce precip stats job. I will see if there is anything more to these unknown LNET errors. In the meantime, I will delve into the other stats jobs.

malloryprow commented:
Noted with the ecme and naefs precip. I'll fix that.

What are these LNET errors? I searched "LNET" in /lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_global_ens_cmce_atmos_precip_stats.o93093411 and I'm not getting anything back.

StevenSimon-NOAA commented:
> /lfs/h2/emc/ptmp/shelley.melchior/EVS_out/

Try a search with just "Error" for both cmce and gefs precip stats. It should appear towards the bottom of both logfiles.

malloryprow committed:
Fix bugs

malloryprow commented:
That isn't related to anything EVS or METplus pretty sure that is something with WCOSS2 nodes.

malloryprow commented:
@ShelleyMelchior-NOAA Can you please rerun dev/drivers/scripts/stats/global_ens/jevs_global_ens_cmce_atmos_precip_stats.sh, dev/drivers/scripts/stats/global_ens/jevs_global_ens_ecme_atmos_precip_stats.sh, dev/drivers/scripts/stats/global_ens/jevs_global_ens_gefs_atmos_precip_stats.sh, and dev/drivers/scripts/stats/global_ens/jevs_global_ens_naefs_atmos_precip_stats.sh?

ShelleyMelchior-NOAA commented:
All previous stats processing has completed.

I have re-submitted the 4 precip drivers noted.
logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/`
`jevs_global_ens_gefs_atmos_precip_stats.o93101107`
`jevs_global_ens_naefs_atmos_precip_stats.o93101121`
`jevs_global_ens_ecme_atmos_precip_stats.o93101037`
`jevs_global_ens_cmce_atmos_precip_stats.o93101509`

malloryprow commented:
jevs_global_ens_ecme_atmos_precip_stats.o93101037 completed and this run does not have the previously mentioned error.

malloryprow commented:
All jobs are completed. I ran my script to grep for "bad" words on the files. The only concerning ones I found were for the previously run precip jobs but the new logs don't have those errors.

ShelleyMelchior-NOAA commented:
This is excellent! Does that seem to satisfy iteration 3 testing?

malloryprow commented:
I looked over the large stat files. I think iteration 3 is good.

Should I remove iteration 4 from this PR since there is https://github.com/NOAA-EMC/EVS/pull/348?

AliciaBentley-NOAA commented:
@malloryprow I would say yes! If this PR gets merged, @MarcelCaron-NOAA can bring your changes into his and be super mindful of any conflicts! 

Does that sound good, @ShelleyMelchior-NOAA?

StevenSimon-NOAA commented:
I can also confirm that all large global_ens stats files are fully accounted for. I am going over the small stats files right now.

ShelleyMelchior-NOAA commented:
Excellent! Yes, iteration 4 will get covered in Marcel's PR #348.
I will continue the formal code review. It will take a little bit of time b/c of how many files were modified. That will allow Steve's small stats file review to continue.

StevenSimon-NOAA commented:
So here's the deal with the global_ens small stats files. In most cases, the counted quantities are less than the expected quantities for each field by model. But the log files indicate no apparent errors. So my conjecture right now is that these discrepancies are due to the recent WCOSS2 issues causing stops and discontinuities in the prep space. Does that make sense in this situation?

malloryprow commented:
I added file checks so we don't see any errors anymore when there is missing data. I added a check too to make sure all the ensemble members are present.

StevenSimon-NOAA commented:
> I added file checks so we don't see any errors anymore when there is missing data. I added a check too to make sure all the ensemble members are present.

In that case, I think we are in good shape with the global_ens small stats files as well.

ShelleyMelchior-NOAA reviewed (approved):
Tests yield expected results. This PR is approved for merge.

ShelleyMelchior-NOAA commented:
Before I merge, I just want to comment, WOW! You did a ton of work on this Mallory. I mean, I know you did a lot of work, but seeing it in lines of changed code in the formal code review is quite staggering. Well done! Thank you for addressing these critical EE2 items for the global_ens component.

StevenSimon-NOAA commented:
> Before I merge, I just want to comment, WOW! You did a ton of work on this Mallory. I mean, I know you did a lot of work, but seeing it in lines of changed code in the formal code review is quite staggering. Well done! Thank you for addressing these critical EE2 items for the global_ens component.

I second these kudos. These updates to the global_ens component have already helped me better navigate the global_ens logfiles, especially those pesky "red herring errors" that pop up from time to time. Many thanks Mallory for the time and effort you put into these updates! We are now one step closer (pun intended) to getting this EVS component ready for production. Onto global_ens plots!

malloryprow commented:
Thank you @ShelleyMelchior-NOAA and @StevenSimon-NOAA!
